src/odeproblem.cpp

```diff
@@ -30,17 +30,31 @@
 static Rcpp::NumericMatrix OMEGADEF(1,1);
 static arma::mat OMGADEF(1,1,arma::fill::zeros);
 
-void dosimeta(void* prob_) {
+void dosimeta(void* prob_, int n) {
```

Comment:
I feel overloading makes more sense than default args here as the two scenarios follow different logic:
```cpp
void dosimeta(void* prob_) {
  odeproblem* prob = reinterpret_cast<odeproblem*>(prob_);
  arma::mat eta = prob->mv_omega(1);
  for(unsigned int i=0; i < eta.n_cols; ++i) {
    prob->eta(i,eta(0,i)); 
  }
}
```
and
```cpp
void dosimeta(void* prob_, int n) {
  odeproblem* prob = reinterpret_cast<odeproblem*>(prob_);
  arma::mat eta = prob->mv_omega(1);
  if(n > int(eta.n_cols)) {
    throw Rcpp::exception("simeta index out of bounds", false);
  }
// is it necessary to throw when n < 0?
  if(n > 0) {
    prob->eta(n-1,eta(0,n-1));
  }
}
```
and `resim` calls the two functions as
```cpp
struct resim {
  //! resim constructor
  resim(refun x, void* y) : fun(x), prob(y){}
  resim(){}
  void operator()() {
    return fun(prob);
  }
  void operator()(int n) {
    return fun(prob, n);
  }
// ...
```

Reply:
Thanks, @yizhang-yiz ; I did try this approach, but I wonder if this is getting hung up on the fact that `fun` is a function pointer?

```c++
typedef void (*refun)(void*, int n);
```

So when I called `fun(prob)` in the functor,  it complained that there was only one argument. 

For this to work would I need to keep pointers to both versions and call the right one?

```c++
typedef void (*refun1)(void*);

typedef void (*refun2)(void*, int n);
```

```

Reply:
yes, we need to `tyedef` two function pointers. I forgot that:
```cpp
struct resim {
   refun1 fun1;
   refun2 fun2;
   void* prob;

  //! resim constructor
  resim(refun1 x, void* y) : fun1(x), prob(y){}
  resim(refun2 x, void* y) : fun2(x), prob(y){}
  resim(){}
  void operator()() {
    return fun1(prob);
  }
  void operator()(int n) {
    return fun2(prob, n);
  }
```
But this is actually where function pointer become cumbersome. Instead we can consider template
```cpp
// define two functors
struct dosimeta {
 void operator()(void* prob_, int n) {...}
 void operator()(void* prob_) {...}
};
struct dosimeps {
 void operator()(void* prob_, int n) {...}
 void operator()(void* prob_) {...}
};
```
and 
```cpp
template<typename F>
struct resim {
  //! resim constructor
  resim(F& x, void* y) : fun(x), prob(y){}
  resim(){}
  void operator()() {
    return fun(prob);
  }
  void operator()(int n) {
    return fun(prob, n);
  }
```